collectors/ebpf.plugin/ebpf_oomkill.c

```diff
@@ -0,0 +1,198 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+#include "ebpf.h"
+#include "ebpf_oomkill.h"
+
+struct config oomkill_config = { .first_section = NULL,
+    .last_section = NULL,
+    .mutex = NETDATA_MUTEX_INITIALIZER,
+    .index = { .avl_tree = { .root = NULL, .compar = appconfig_section_compare },
+        .rwlock = AVL_LOCK_INITIALIZER } };
+
+#define OOMKILL_MAP_KILLCNT 0
+static ebpf_local_maps_t oomkill_maps[] = {
+    {
+        .name = "tbl_oomkill",
+        .internal_input = NETDATA_OOMKILL_MAX_ENTRIES,
+        .user_input = 0,
+        .type = NETDATA_EBPF_MAP_STATIC,
+        .map_fd = ND_EBPF_MAP_FD_NOT_INITIALIZED
+    },
+    /* end */
+    {
+        .name = NULL,
+        .internal_input = 0,
+        .user_input = 0,
+        .type = NETDATA_EBPF_MAP_CONTROLLER,
+        .map_fd = ND_EBPF_MAP_FD_NOT_INITIALIZED
+    }
+};
+
+static ebpf_data_t oomkill_data;
+
+static ebpf_tracepoint_t oomkill_tracepoints[] = {
+    {.enabled = false, .class = "oom", .event = "mark_victim"},
+    /* end */
+    {.enabled = false, .class = NULL, .event = NULL}
+};
+
+static struct bpf_link **probe_links = NULL;
+static struct bpf_object *objects = NULL;
+
+/**
+ * Clean up the main thread.
+ *
+ * @param ptr thread data.
+ */
+static void oomkill_cleanup(void *ptr)
+{
+    ebpf_module_t *em = (ebpf_module_t *)ptr;
+    if (!em->enabled) {
+        return;
+    }
+
+    if (probe_links) {
+        struct bpf_program *prog;
+        size_t i = 0 ;
+        bpf_object__for_each_program(prog, objects) {
+            bpf_link__destroy(probe_links[i]);
+            i++;
+        }
+        bpf_object__close(objects);
+    }
+}
+
+static void oomkill_write_data()
+{
+    // the first `i` entries of `keys` will contain the currently active PIDs
+    // in the eBPF map.
+    uint32_t i = 0;
+    int32_t keys[NETDATA_OOMKILL_MAX_ENTRIES] = {0};
+
+    uint32_t curr_key = 0;
+    uint32_t key = 0;
+    int mapfd = oomkill_maps[OOMKILL_MAP_KILLCNT].map_fd;
+    while (bpf_map_get_next_key(mapfd, &curr_key, &key) == 0) {
+        curr_key = key;
+
+        keys[i] = key;
+        i += 1;
+
+        // delete this key now that we've recorded its existence. there's no
+        // race here, as the same PID will only get OOM killed once.
+        int test = bpf_map_delete_elem(mapfd, &key);
+        if (unlikely(test < 0)) {
+            // since there's only 1 thread doing these deletions, it should be
+            // impossible to get this condition.
+            error("key unexpectedly not available for deletion.");
+        }
+    }
+
+    // for each app, see if it was OOM killed. record as 1 if so otherwise 0.
+    struct target *w;
+    for (w = apps_groups_root_target; w != NULL; w = w->next) {
+        if (likely(w->exposed && w->processes)) {
+            bool was_oomkilled = false;
```

Comment:
I do not have objections with usage of `bool`, but I do not remember functions where we are using this data type, I will suggest you use `CONFIG_BOOLEAN_YES` and `CONFIG_BOOLEAN_NO` to keep pattern with the rest of the code.

Reply:
I think "CONFIG" prefix indicates that it is only for configuration.

<details>
<summary>grep "bool " --include "*.c" --include "*.h" -R *</summary>

```cmd
collectors/ebpf.plugin/ebpf.h:    bool enabled;
collectors/ebpf.plugin/ebpf_hardirq.c:        bool v_is_new = false;
collectors/ebpf.plugin/ebpf_hardirq.c:        bool name_saved = false;
collectors/ebpf.plugin/ebpf_hardirq.h:    bool dim_exists; // keep this after `int irq` for alignment byte savings.
collectors/proc.plugin/proc_net_wireless.c:    bool configured;
exporting/tests/test_exporting_engine.h:bool __wrap_mongoc_uri_set_option_as_int32(const mongoc_uri_t *uri, const char *option, int32_t value);
exporting/tests/test_exporting_engine.h:bool __wrap_mongoc_client_set_appname(mongoc_client_t *client, const char *appname);
exporting/tests/test_exporting_engine.h:bool __wrap_mongoc_collection_insert_many(
exporting/tests/exporting_doubles.c:bool __wrap_mongoc_uri_set_option_as_int32 (const mongoc_uri_t *uri, const char *option, int32_t value)
exporting/tests/exporting_doubles.c:bool __wrap_mongoc_client_set_appname (mongoc_client_t *client, const char *appname)
exporting/tests/exporting_doubles.c:bool __wrap_mongoc_collection_insert_many(
libnetdata/ebpf/ebpf.h:    bool retprobe;
web/api/tests/web_api.c:static void build_request(struct web_buffer *wb, const char *url, bool use_cr, size_t num_headers)
web/api/tests/web_api.c:    bool completed, use_cr;
```

</details>

Reply:
Right, `bool` is a very standard data type so I don't see any issue with that, even if it isn't used that much already. It better represents the data we're trying to store here.